### ISSUE TITLE
Imagenet evaluation error

### ISSUE THREAD

**Issue by chs20** [mlfoundations/open_flamingo#132]
When running
```
python open_flamingo/eval/evaluate.py \
--lm_path llama/7B-hf \
--lm_tokenizer_path llama/7B-hf \
--checkpoint_path OpenFlamingo-9B/checkpoint.pt \
--imagenet_root datasets/imagenet \
--eval_imagenet \
--cross_attn_every_n_layers 2 \
--device 1
```
I get the following error:
```
Using pad_token, but it is not set yet.
Loading checkpoint shards: 100%|
Flamingo model initialized with 2425735200 trainable parameters
Evaluating on ImageNet...
processing batch 0 of 625
  0%|                                                                                                                                                                                                                                          | 0/1000 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "openflamingo/open_flamingo/eval/evaluate.py", line 961, in <module>
    main()
  File "openflamingo/open_flamingo/eval/evaluate.py", line 295, in main
    imagenet_score = evaluate_imagenet(
  File "openflamingo/open_flamingo/eval/evaluate.py", line 917, in evaluate_imagenet
    per_sample_probs = compute_per_sample_probs(
  File "openflamingo/open_flamingo/eval/classification.py", line 76, in compute_per_sample_probs
    shift_logits, shift_labels = compute_shifted_logits_and_labels(
  File "openflamingo/open_flamingo/eval/classification.py", line 57, in compute_shifted_logits_and_labels
    end_of_prefix = -labels[idx].tolist()[::-1].index(tokenizer.eos_token_id) - 1
ValueError: 2 is not in list
```

Any ideas how to fix it?

**Comment by anas-awadalla** [comment]
hi! Thanks for pointing this out, I am taking a closer look. By the name of your model directory I am guessing you are using this [version](https://huggingface.co/decapoda-research/llama-7b-hf) of llama? I was wondering if this has to do with the [unspecified EOS/BOS tokens](https://huggingface.co/decapoda-research/llama-7b-hf/blob/main/tokenizer_config.json) in the tokenizer config but I am not certain yet.

Also something else I want to point out. You should pass in 4 for cross_attn_every_n_layers argument if you are using the checkpoint we provide.

**Comment by chs20** [comment]
Hi, thanks for your reply :) I converted the llama weights to HF format with the script linked in the readme. However, the tokens in `tokenizer_config.json` are indeed unspecified.

Thanks for the hint on the cross_attn_every_n_layers argument. The default value in `evaluate.py` is `1`, but it did not fit on my GPU, so I increased it.

**Comment by jpgard** [comment]
Howdy! Thanks for using OF.

Here is an example of a working command to run the eval; if this doesn't work there may be some differences in your setup that we can't necessarily debug for you:

```
python open_flamingo/eval/evaluate.py \
    --lm_path facebook/opt-1.3b \
    --lm_tokenizer_path facebook/opt-1.3b \
    --clip_path openai/clip-vit-large-patch14 \
    --checkpoint_path/path/to/checkpoint.pt \
    --device 0 \
    --imagenet_root /tmp/val \
    --eval_imagenet \
    --results_file tmp.json \
    --num_samples 8 --shots 8 --num_trials 1
```

You'll need to change the `lm_path` to your local LLaMa weights if you would like to use LLaMa.

I want to emphasize that **there are still some bugs in the ImageNet evals** that we are working out -- so, while this will help you debug your setup, I wouldn't trust the numbers it produces quite yet :). We'll be updating with an improved (and faster) version of the classification evals soon.

**Comment by jpgard** [comment]
(specifically I don't think you need to specify `cross_attn_every_n_layers`)

**Comment by chs20** [comment]
Hi @jpgard :) It seems that the argument `--clip_path` does not exist for `evaluate.py` - I get `evaluate.py: error: unrecognized arguments: --clip_path`. Is `--vision_encoder_path` the correct one? 

However, I get the same error as in my initial post. As @anas-awadalla suspects, it's probably due to the tokenizer config.

**Comment by jpgard** [comment]
>  Is --vision_encoder_path the correct one?
I think so -- @anas-awadalla can correct if not. This must have changed recently, sorry :)

I can't advise on the tokenizer config, will let Anas chime in there.

**Comment by anas-awadalla** [comment]
@chs20 Yeah the clip args changed very recently. The default ones should be fine to use.

**Comment by chs20** [comment]
When setting the tokens in `tokenizer_config.json` manually, it runs :)
```
{"bos_token": "<s>", "eos_token": "</s>", "model_max_length": 1000000000000000019884624838656, "tokenizer_class": "LlamaTokenizer", "unk_token": "<unk>"}
```

**Comment by vtddggg** [comment]
> Howdy! Thanks for using OF.
> 
> Here is an example of a working command to run the eval; if this doesn't work there may be some differences in your setup that we can't necessarily debug for you:
> 
> ```
> python open_flamingo/eval/evaluate.py \
>     --lm_path facebook/opt-1.3b \
>     --lm_tokenizer_path facebook/opt-1.3b \
>     --clip_path openai/clip-vit-large-patch14 \
>     --checkpoint_path/path/to/checkpoint.pt \
>     --device 0 \
>     --imagenet_root /tmp/val \
>     --eval_imagenet \
>     --results_file tmp.json \
>     --num_samples 8 --shots 8 --num_trials 1
> ```
> 
> You'll need to change the `lm_path` to your local LLaMa weights if you would like to use LLaMa (and specify `cross_attn_every_n_layers ` as @anas-awadalla mentions above).
> 
> I want to emphasize that **there are still some bugs in the ImageNet evals** that we are working out -- so, while this will help you debug your setup, I wouldn't trust the numbers it produces quite yet :). We'll be updating with an improved (and faster) version of the classification evals soon.

@jpgard I try this command, however got size mismatch error. 
Does the [OpenFlamingo 9B model](https://huggingface.co/openflamingo/OpenFlamingo-9B) model be fully compatible with OPT?

**Comment by chs20** [comment]
@vtddggg When I run
```
python open_flamingo/eval/evaluate.py \
--lm_path llama/7B-hf \
--lm_tokenizer_path llama/7B-hf \
--checkpoint_path OpenFlamingo-9B/checkpoint.pt \
--imagenet_root datasets/imagenet \
--eval_imagenet \
--cross_attn_every_n_layers 4 \
--shots 0 4 8 \
--num_samples 8 \
--batch_size 1 \
--results_file results/eval_results.json \
--device 1
```
the results file reads
```
{"imagenet": [{"shots": 0, "trials": [0.75], "mean": 0.75}, {"shots": 4, "trials": [1.0], "mean": 1.0}, {"shots": 8, "trials": [1.0], "mean": 1.0}]}
```

**Comment by vtddggg** [comment]
@chs20 I follow the standard step in README, and modify the `tokenizer_config.json`

> When setting the tokens in `tokenizer_config.json` manually, it runs :)
> 
> ```
> {"bos_token": "<s>", "eos_token": "</s>", "model_max_length": 1000000000000000019884624838656, "tokenizer_class": "LlamaTokenizer", "unk_token": "<unk>"}
> ```

But still got `ImageNet score: 0.0`. Is there something else I missed?😂
I will be appreciate if you can help me to check the output of [example](https://github.com/mlfoundations/open_flamingo#generating-text) in README.md. For me, the output is 
```
Generated text:    <s> <image>  An image of two cats. <|endofchunk|> <image>  An image of a bathroom sink. <|endofchunk|> <image>  An image of a Thanksgiving table. An image of a Christmas tree. An image of a Christmas tree.
```

**Comment by anas-awadalla** [comment]
@vtddggg As @jpgard [mentioned](https://github.com/mlfoundations/open_flamingo/issues/132#issuecomment-1489184247) there are some bugs we noticed in imagenet evals that we are fixing. As for your output from the example, that looks fine to me. Check out this discussion on the [repetition](https://github.com/mlfoundations/open_flamingo/issues/147) bug.

**Comment by vishaal27** [comment]
Hey, just checking in if the ImageNet eval bugs mentioned previously have been fixed?

**Comment by vtddggg** [comment]
@vishaal27 Have you encountered similar problems in https://github.com/mlfoundations/open_flamingo/issues/132#issuecomment-1491616076? Afterwards, I checked the imagenet eval codes, and find something strange in: https://github.com/mlfoundations/open_flamingo/blob/e6cbff8d55469f75fd1fc564a3eda5f5a44d01de/open_flamingo/eval/evaluate.py#L864 In Flamingo, LLM is conditioned on input images. However, in this implementation, `context_precomputed` seems invariant even input conditioned images are changed. It makes the model output same results on different class names. 

@anas-awadalla I'm still finding where causing this problem. Can you make some check to `print(context_precomputed)` during evaluation, to verify that this issue is not specific to me

**Comment by anas-awadalla** [comment]
cc @jpgard

**Comment by vtddggg** [comment]
Till now I still cannot find where causing the crash on ImageNet evaluation. 

Instead, I can provide a simple ImageNet evaluation script which is written by myself but is more intuitive than implementation in [evaluate.py](https://github.com/mlfoundations/open_flamingo/blob/bee358f445c55ca3297c5563da25b6285814234a/open_flamingo/eval/evaluate.py): 
```python
from open_flamingo import create_model_and_transforms
from tqdm import tqdm
from open_flamingo.eval.imagenet_utils import (
    openai_imagenet_classnames,
    IMAGENET_1K_CLASS_ID_TO_LABEL,
)
import os
import numpy as np
import torch
import random
from PIL import Image

class ImageNetDataset(ImageFolder):
    """Class to represent the ImageNet1k dataset."""

    def __init__(self, root, **kwargs):
        super().__init__(root=root, **kwargs)

    def __getitem__(self, idx):
        sample, target = super().__getitem__(idx)
        target_label = IMAGENET_1K_CLASS_ID_TO_LABEL[target]
        return {
            "image": sample,
            "class_id": target,  # numeric ID of the ImageNet class
            "class_name": target_label,  # human-readable name of ImageNet class
        }

def find_sub_list(sl,l):
    results=[]
    sll=len(sl)
    for ind in (i for i,e in enumerate(l) if e==sl[0]):
        if l[ind:ind+sll]==sl:
            results.append(ind+sll-1)
    return results

model, image_processor, tokenizer = create_model_and_transforms(
    clip_vision_encoder_path="ViT-L-14",
    clip_vision_encoder_pretrained="openai",
    lang_encoder_path="/path/to/hf_llama_7b",
    tokenizer_path="/path/to/hf_llama_7b",
    cross_attn_every_n_layers=4
)

checkpoint_path = '/path/to/checkpoint.pt'
model.load_state_dict(torch.load(checkpoint_path), strict=False)
model.cuda()

train_dataset = ImageNetDataset('/path/to/imagenet/train')
val_dataset = ImageNetDataset('/path/to/imagenet/val')

context_num = 4

# random context samples
random_indices = np.random.choice(
        len(train_dataset), context_num, replace=False
    )
in_context_samples = [train_dataset[i] for i in random_indices]

acc1 = 0
acc5 = 0
count = 0
for i, batch in enumerate(val_dataset):

    vision_x = [image_processor(data['image']).unsqueeze(0) for data in in_context_samples] + [image_processor(batch['image']).unsqueeze(0)]
    vision_x = torch.cat(vision_x, dim=0)
    vision_x = vision_x.unsqueeze(1).unsqueeze(0)

    overall_probs = []
    for imagenet_class_name in openai_imagenet_classnames:

        tokenizer.padding_side = "left" # For generation padding tokens should be on the left
        lang_x = tokenizer(
            ["<image>A photo of a {}<|endofchunk|><image>A photo of a {}<|endofchunk|><image>A photo of a {}<|endofchunk|><image>A photo of a {}<|endofchunk|><image>A photo of a {}".format(in_context_samples[0]['class_name'], in_context_samples[1]['class_name'], in_context_samples[2]['class_name'], in_context_samples[3]['class_name'], imagenet_class_name)],
            return_tensors="pt",
        )

        outputs = model(
            vision_x=vision_x.cuda(),
            lang_x=lang_x["input_ids"].cuda(),
            attention_mask=lang_x["attention_mask"].cuda(),
            clear_conditioned_layers=False
        )
        probs = torch.softmax(outputs.logits, dim=-1).detach()
        # collect the probability of the generated token -- probability at index 0 corresponds to the token at index 1
        probs = probs[:, :-1, :]
        input_ids = lang_x["input_ids"][:, 1:].cuda()
        gen_probs = torch.gather(probs, 2, input_ids[:, :, None]).squeeze(-1)

        probs = []
        for input_sentence, input_probs in zip(input_ids, gen_probs):
            text_sequence = []
            idxes = find_sub_list([32001, 319, 15373, 310, 263], input_sentence.detach().cpu().numpy().tolist())
            input_sentence = input_sentence[idxes[-1]+1:]
            input_probs = input_probs[idxes[-1]+1:]
            probs.append(torch.prod(input_probs).item())
        overall_probs.append(probs)

    count += 1
    top5 = [IMAGENET_1K_CLASS_ID_TO_LABEL[pred] for pred in np.argsort(np.array(overall_probs)[:,0])[::-1][:5]]
    if batch['class_name'] == top5[0]:
        acc1 += 1
    if batch['class_name'] in top5:
        acc5 += 1
    print('eval {}/{}: acc@1 ({}), acc@5 ({})'.format(i, len(val_dataset), acc1/count, acc5/count))
```

Wish this code can help someone who want to evaluate flamingo on ImageNet! 
I'm not sure that the script is completely correct. If there are any problem, feel free to comment in this issue. 
Note that context text input is hard coded. Remember to modify it when you change `context_num`.

**Comment by jpgard** [comment]
Hi @vtddggg , just wanted to say sorry for the slow reply and also thanks for noticing all this. Just submitted a PR to fix this issue, once #185 is merged these issues should be fixed. You will see some code that looks familiar in there :) thank you for your discussion here!!

fwiw I think the problem was actually that when using cached activations (passing `past_key_values` and `use_cached=True`), decoding has to be done **one token at a time**, and we were only decoding the first token. Our previous OPT-based models weren't nearly as accurate so we didn't notice the regression when this bug was introduced!

We'll reintroduce text activation caching in a future PR, but for now we wanted to get a slightly less efficient, but correct, evaluation loop implemented :). A huge amount of compute is wasted recomputing image and text features for the imagenet evals (since only the last few tokens ever change within a batch), so eliminating all of that redundant computation has been and still is a top priority for us.
